bot/cogs/moderation/incidents.py

```diff
@@ -0,0 +1,331 @@
+            log.trace(f"Skipping emoji as it's already been placed: {signal_emoji}")
+        else:
+            log.trace(f"Adding reaction: {signal_emoji}")
+            await incident.add_reaction(signal_emoji.value)
+
+
+class Incidents(Cog):
+    """
+    Automation for the #incidents channel.
+
+    This cog does not provide a command API, it only reacts to the following events.
+
+    On start-up:
+        * Crawl #incidents and add missing `Signal` emoji where appropriate
+        * This is to retro-actively add the available options for messages which
+          were sent while the bot wasn't listening
+        * Pinned messages and message starting with # do not qualify as incidents
+        * See: `crawl_incidents`
+
+    On message:
+        * Add `Signal` member emoji if message qualifies as an incident
+        * Ignore messages starting with #
+            * Use this if verbal communication is necessary
+            * Each such message must be deleted manually once appropriate
+        * See: `on_message`
+
+    On reaction:
+        * Remove reaction if not permitted (`ALLOWED_EMOJI`, `ALLOWED_ROLES`)
+        * If `Signal.ACTIONED` or `Signal.NOT_ACTIONED` were chosen, attempt to
+          relay the incident message to #incidents-archive
+        * If relay successful, delete original message
+        * See: `on_raw_reaction_add`
+
+    Please refer to function docstrings for implementation details.
+    """
+
+    def __init__(self, bot: Bot) -> None:
+        """Prepare `event_lock` and schedule `crawl_task` on start-up."""
+        self.bot = bot
+
+        self.event_lock = asyncio.Lock()
+        self.crawl_task = self.bot.loop.create_task(self.crawl_incidents())
+
+    async def crawl_incidents(self) -> None:
+        """
+        Crawl #incidents and add missing emoji where necessary.
+
+        This is to catch-up should an incident be reported while the bot wasn't listening.
+        After adding each reaction, we take a short break to avoid drowning in ratelimits.
+
+        Once this task is scheduled, listeners that change messages should await it.
+        The crawl assumes that the channel history doesn't change as we go over it.
+        """
+        await self.bot.wait_until_guild_available()
+        incidents: discord.TextChannel = self.bot.get_channel(Channels.incidents)
+
+        # Limit the query at 50 as in practice, there should never be this many messages,
+        # and if there are, something has likely gone very wrong
+        limit = 50
+
+        # Seconds to sleep after adding reactions to a message
+        sleep = 2
+
+        log.debug(f"Crawling messages in #incidents: {limit=}, {sleep=}")
+        async for message in incidents.history(limit=limit):
+
+            if not is_incident(message):
+                log.trace("Skipping message: not an incident")
+                continue
+
+            if has_signals(message):
+                log.trace("Skipping message: already has all signals")
```

Comment:
Yes, you're right, these should log the id. I reviewed my logs and added ids where I felt it was appropriate in 20b27f3.

It's an interesting problem because if you have a stream of `trace`/`debug` logs sprinkled throughout a function, you don't want to be logging the ids in each message. So I usually just log this information in the first one, and then it's clear that "message" refers to the same object. But if they are followed by an >= `info` log, it should probably log the id again, because we usually have logs below that disabled in production. It's also fairly easy to forget to adjust other logs when changing just one.

Hopefully I got it more right this time, let me know what you think.